### DISCUSSION TITLE
Unit conversions for collections

### DISCUSSION THREAD

**Discussion by md-amf1** [angularsen/UnitsNet#913] [Q&A] [answered]
Hi @angularsen,

I've just been playing with UnitsNet and it looks like a really useful library.

I'm attempting to use it to convert collections of values stored in any source unit to SI units.

I've been looking through the library and didn't spot any obvious solutions for converting collections rather than individual values, I may well have missed something though.

An example of the approach I've used to try this out simply to loop through the source array generating elementwise conversions, e.g.
```
public double[] ToSI(double[] values, LengthUnit unit)
{
    var results = new double[values.Length];
    for (var i = 0; i < results.Length; i++)
    {
        results[i] = ToSI(values[i], unit);
    }
    return results;
}

public double ToSI(double value, LengthUnit unit)
{
    return Length.From(value, unit).ToUnit(UnitSystem.SI).Value;
}
```

Before I go fully down this route, I just wondered if there was a pre-baked solution for converting values in arrays/collections .

Thanks!

**Comment by angularsen** [comment, marked as answer]
Hi, there is nothing in the library for collections specifically, but your approach works and you can also consider using LINQ to simplify slightly:
```c#
// LINQPad
void Main()
{
	ToSI(new double[] { 1, 2, 3}, LengthUnit.Centimeter).Dump();
}

public double[] ToSI(double[] values, LengthUnit unit)
{
	return values.Select(val => Length.From(val, unit).ToUnit(UnitSystem.SI).Value).ToArray();
}
```

**Reply by md-amf1** [reply]
Thanks, I'm glad I wasn't missing anything.

Due to not knowing which source units I'm ultimately expecting, I generated the following which does the trick quite nicely, going via `Enum` rather than each specific unit type so I can use the same method for LengthUnit, AreaUnit etc.

```
public double[] ToSI(double[] values, Enum unit)
{
    var results = new double[values.Length];
    for (var i = 0; i < results.Length; i++)
    {
        results[i] = ToSI(values[i], unit);
    }
    return results;
}

public double ToSI(double value, Enum unit)
{
    return unit switch
    {
        AccelerationUnit acceleration => ToSI(value, acceleration, Acceleration.From),
        AmountOfSubstanceUnit amountOfSubstance => ToSI(value, amountOfSubstance, AmountOfSubstance.From),
        AmplitudeRatioUnit amplitudeRatio => ToSI(value, amplitudeRatio, AmplitudeRatio.From),
        AngleUnit angle => ToSI(value, angle, Angle.From),
        // etc.
    }
}

private static double ToSI<TUnit, TQuantity>(
    double value,
    TUnit unit,
    Func<QuantityValue, TUnit, TQuantity> from)
    where TQuantity: IQuantity
{
    return from(value, unit).ToUnit(UnitSystem.SI).Value;
}

**Reply by angularsen** [reply]
Glad you worked it out!
